### ISSUE TITLE
Memory module with nested Sequencer

### ISSUE THREAD

**Issue by vzhong** [Element-Research/rnn#224]
Hi,

I was wondering if someone could give me some pointers on how to implement a memory module that is as follows:

- takes an 3d input tensor `(batch, slices, embeddings)`
- examines the input tensor `T` times
- each time step, the module
  - computes an attention (Eg. weighted average of the input slices in which the weights are computed using the input at slice `i` and the previous memory)
  - computes the next memory by applying a transformation to the attention

I've attempted an implementation here:

https://gist.github.com/vzhong/63c2971bac7b78d3ea64180495fcece1

I've faked the "examine the input `T` times" with a call to `Sequencer` with a table that contains `T` copies of the input.

This network `forward`s just fine, but does not `backward` (eg. can't compute its gradients). I suspect I'm not using either `Sequencer` or `Recurrent` correctly. The output I get is:

```text
(1,.,.) =
  0.2442  0.7207  0.3475  0.7180
  0.3911  0.7244  0.7628  0.8934
  0.1389  0.3448  0.5133  0.5745

(2,.,.) =
  0.7835  0.9123  0.0191  0.5046
  0.1511  0.3550  0.4622  0.4324
  0.7526  0.3369  0.5883  0.2316
[torch.DoubleTensor of size 2x3x4]

recurrent memory module
{
  1 : DoubleTensor - size: 2x3x4
  2 : DoubleTensor - size: 2x3x4
  3 : DoubleTensor - size: 2x3x4
}
{
  1 : DoubleTensor - size: 2x20
  2 : DoubleTensor - size: 2x20
  3 : DoubleTensor - size: 2x20
}
/Users/victor/torch/install/bin/luajit: /Users/victor/torch/install/share/lua/5.1/nn/Container.lua:69:
In 2 module of nn.Sequential:
/Users/victor/torch/install/share/lua/5.1/rnn/Recursor.lua:41: assertion failed!
stack traceback:
        [C]: in function 'assert'
        /Users/victor/torch/install/share/lua/5.1/rnn/Recursor.lua:41: in function '_updateGradInput'
        ...or/torch/install/share/lua/5.1/rnn/AbstractRecurrent.lua:46: in function 'updateGradInput'
        /Users/victor/torch/install/share/lua/5.1/rnn/Sequencer.lua:78: in function </Users/victor/torch/install/share/lua/5.1/rnn/Sequencer.lua:71>
        [C]: in function 'xpcall'
        /Users/victor/torch/install/share/lua/5.1/nn/Container.lua:65: in function 'rethrowErrors'
        /Users/victor/torch/install/share/lua/5.1/nn/Sequential.lua:55: in function 'updateGradInput'
        ...s/victor/torch/install/share/lua/5.1
```

The prints are of the inputs and the result of the `forward` call. The crash occurs during `backward`.

Thanks for your help!

**Comment by nicholas-leonard** [comment]
@vzhong This doesn't crash : 
```lua
require 'nn'
require 'dpnn'
require 'rnn'
require 'nngraph'

local opt = {
  n_seq = 3,
  d_hid = 4,
  d_mem = 20,
  n_batch = 2,
}

local build_memory_module = function()
  local x = nn.Identity()()
  local mem_tm1 = nn.Identity()()
  local input = {x, mem_tm1}

  local x_list = nn.SplitTable(2)(x)

  -- simple scorer that produces a score s_i given {mem_tm1, x_i}
  local scorer = nn.Sequential()
      :add(nn.JoinTable(2))
      :add(nn.Linear(opt.d_hid + opt.d_mem, 1))
  -- produce a table of scores for each slice in the input
  local seq1 = nn.Sequencer(scorer)
  seq1:remember()
  local scores = nn.Sequential()
      :add(nn.ZipTableOneToMany())
      :add(seq1){mem_tm1, x_list}
  -- normalize scores via softmax
  local normalized_scores = nn.Sequential()
      :add(nn.JoinTable(2))
      :add(nn.SoftMax())
      :add(nn.SplitTable(2))(scores)
  -- expand scores for multiplcation
  local seq2 = nn.Sequencer(nn.Replicate(opt.d_hid, 2))
  seq2:remember()
  local expanded_scores = seq2(normalized_scores)

  -- multiply each slice of input by corresponding expanded scores and sum
  local seq3 = nn.Sequencer(nn.CMulTable())
  seq3:remember()
  local attn = nn.Sequential()
      :add(nn.ZipTable())
      :add(seq3)
      :add(nn.CAddTable()){expanded_scores, x_list}

  -- compute next state
  local mem = nn.Tanh()(nn.Linear(opt.d_hid, opt.d_mem)(attn))

  local output = {mem}
  return nn.gModule(input, output)
end


local x = torch.rand(opt.n_batch, opt.n_seq, opt.d_hid)
print(x)

local memory_module = build_memory_module()

memory_module = nn.Sequencer(nn.Recurrence(memory_module, opt.d_mem, 2))
print('recurrent memory module')
local inputs = {}
for t = 1, 3 do
  table.insert(inputs, x:clone())
end
print(inputs)
local out_memory = memory_module:forward(inputs)
print(out_memory)

local dout_memory = {}
for t = 1, #out_memory do
  dout_memory[t] = out_memory[t]:clone()
end
memory_module:backward(inputs, dout_memory)
```

Let me know if it works.

**Comment by vzhong** [comment]
Thank you, it works!
